### DISCUSSION TITLE
How do you dynamically import multiple named exports from a module in one go?

### DISCUSSION THREAD

**Discussion by paambaati** [vercel/next.js#13139] [Help]
Say I'm using —

```javascript
import { ComposableMap, Geographies, Geography, Marker } from 'react-simple-maps';
```

If I want to import this dynamically, I'll have to do something like —

```javascript
import dynamic from 'next/dynamic';

const ComposableMap = dynamic(() =>
  import('react-simple-maps').then(module => module.ComposableMap)
);
const Geographies = dynamic(() =>
  import('react-simple-maps').then(module => module.Geographies)
);
const Geography = dynamic(() =>
  import('react-simple-maps').then(module => module.Geography)
);
const Marker = dynamic(() =>
  import('react-simple-maps').then(module => module.Marker)
);
```

Is there a better way to do this? Destructuring doesn't seem to help.

**Comment by VivekNeel** [comment]
May be you can try like this? 

```
const {
ComposableMap, 
Geographies,
...rest
} = dynamic(() =>
  import('react-simple-maps').then(module => module)
);
```

**Reply by paambaati** [reply]
Unfortunately, this doesn't work.

```
Error: Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: undefined. You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.
```

**Reply by surjithctly** [reply]
@paambaati How did you solve this?

**Comment by michielmetcake** [comment]
```const {
ComposableMap, 
Geographies,
...rest
} = {
ComposableMap: dynamic(() =>
  import('react-simple-maps').then(module => module.ComposableMap)
),
Geographies: dynamic(() =>
  import('react-simple-maps').then(module => module.Geographies)
)}
;```

**Reply by obiknows** [reply]
do you know if this increases build size, or is react that smart to only do the `import('react-simple-maps')` once?

**Reply by michielmetcake** [reply]
Not sure about this. But in our case it did increase the bundle size.

**Comment by 1finedev** [comment]
has this been solved?

**Comment by Technorohit** [comment]
You can try this. 
```
import('react-simple-maps').then(module =>
{
const {ComposableMap}= module;
})
```

**Comment by zaaakher** [comment]


**Reply by dsacramone** [reply]
Weird, I expected this to work, but it did not - for me.

**Reply by shanejones** [reply]
Same, did not work for me either

**Comment by alicerocheman** [comment]
> Say I'm using —
> 
> ```js
> import { ComposableMap, Geographies, Geography, Marker } from 'react-simple-maps';
> ```
> 
> If I want to import this dynamically, I'll have to do something like —
> 
> ```js
> import dynamic from 'next/dynamic';
> 
> const ComposableMap = dynamic(() =>
>   import('react-simple-maps').then(module => module.ComposableMap)
> );
> const Geographies = dynamic(() =>
>   import('react-simple-maps').then(module => module.Geographies)
> );
> const Geography = dynamic(() =>
>   import('react-simple-maps').then(module => module.Geography)
> );
> const Marker = dynamic(() =>
>   import('react-simple-maps').then(module => module.Marker)
> );
> ```
> 
> Is there a better way to do this? Destructuring doesn't seem to help.

In my case I get typescript errors if i do this. :/
Haven't found a solution yet.

**Reply by kasir-barati** [reply]
Any luck with type error?

**Comment by hasriza** [comment]
Did anyone try:
`import(<path>).then(
({<module1>,<module2>}) => {<module1>.<action>; 
<module2>.<action>;}
 )`

**Comment by wilsonpage** [comment]
I had to add `FC` to my exported component:

```diff
import { FC } from 'react';

- export const MyComponent = ({ foo }: { foo: number }) => {
+ export const MyComponent: FC<{ foo: number }> = ({ foo }) => {
```

and then import like:

```tsx
const MyComponentDynamic = dynamic(
  () => import('./myModule').then((m) => m.MyComponent)
);
```

**Comment by merthanmerter** [comment]
would using a utility function make it cleaner for you?

```
// /lib/nextImport.ts
function nextDynamic(module: string) {
  return dynamic(() => import('@modules/Home').then((mod: any) => mod[module]) as Promise<React.FC<any>>)
}
```

```
// /pages/index.tsx
const Featured = nextDynamic('Featured')
const BlogPost = nextDynamic('BlogPost')
```

**Comment by pritamguha** [comment]
Any suggestion, on how to import RangePicker from antd DatePicker using next js dynamic import?

in general, I'm using  the following code for getting RangePicker,

```
import {DatePicker} from "antd";
const {RangePicker} = DatePicker;
```

**Reply by merthanmerter** [reply]
> Any suggestion, on how to import RangePicker from antd DatePicker using next js dynamic import?
> 
> in general, I'm using the following code for getting RangePicker,
> 
> ```
> import {DatePicker} from "antd";
> const {RangePicker} = DatePicker;
> ```

You could use RangePicker in a custom component and dynamic import that.

**Reply by bnbon** [reply]
Any update on this?

**Comment by robotic-forest** [comment]
Anyone wandering by trying to import named exports that are not React Components, use [this](https://stackoverflow.com/questions/68323336/next-js-dynamic-import-for-named-export)

**Comment by fredericobreno** [comment]
This worked for me:

```js
const Map = dynamic(
  async () => {
    const { Map } = await import('@/components/Map')
    return { default: Map }
  },
  {
    ssr: false,
    loading: () => <p>Loading...</p>,
  },
)```

**Reply by kasir-barati** [reply]
JFI*: ` Type 'RealmPlugin' is not assignable to type 'ReactNode'.ts(2345)`. Not gonna work when trying to import something that is not a component:
```ts
const headingsPlugin = dynamic(async () => {
  const {headingsPlugin} = await import('@mdxeditor/editor')
  return {default: headingsPlugin}
});
```

*Personally made this mistake :sweat_smile:.

[Learn more](https://nextjs.org/docs/pages/building-your-application/optimizing/lazy-loading)

**Comment by deriveitCreator** [comment]
How about useRef:


```js
function ThisComp(){
  const geographiesRef = useRef(null);
  const geographyRef = useRef(null);
  const markerRef = useRef(null);
  const [dynamicLoadingDone, setDLD] = useState(false);

  const ComposableMap = dynamic(() => import('react-simple-maps').then(module =>{
    geographiesRef.current = module.Geographies;
    geographyRef.current = module.Geography;
    markerRef.current = module.Marker;
    setDLD(true);
    return module.ComposableMap;
  }));

  //If lets say you want to render to geography one only
  if(dynamicLoadingDone) return <p>Loading</p>
  else{
    let GeoComp = geographyRef.current;
    return <GeoComp/>
  }
}
```

**Reply by deriveitCreator** [reply]
I'm not an expert, so if this is a bad practice, please let me know any alternatives
